File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/ml/processing/PackerRegistry.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.ml.processing;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import java.io.IOException;
import java.util.HashMap;
import org.json.JSONException;
import org.json.JSONObject;
import org.pytorch.IValue;

public class PackerRegistry {

  private static final HashMap<String, Packer> mPackerMap = new HashMap<>();
  private static final HashMap<String, Unpacker> mUnpackerMap = new HashMap<>();

  private PackerRegistry() {}

  public static synchronized void register(final String type, final Packer packer) {
    if (mPackerMap.containsKey(type)) {
      throw new IllegalStateException("Packer for type '" + type + "' is already registered");
    }

    mPackerMap.put(type, packer);
  }

  public static synchronized void register(final String type, final Unpacker unpacker) {
    if (mUnpackerMap.containsKey(type)) {
      throw new IllegalStateException("Unpacker for type '" + type + "' is already registered");
    }

    mUnpackerMap.put(type, unpacker);
  }

  public static IValue pack(
      final JSONObject jobject, final ReadableMap params, final PackerContext packerContext)
      throws JSONException, IOException {
    final String type = jobject.getString("type");
    final Packer p = mPackerMap.get(type);
    if (p == null) {
      throw new IllegalStateException("Unknown type for packer: '" + type + "'");
    }
    return p.pack(jobject, params, packerContext);
  }

  public static void unpack(
      final IValue ivalue,
      final JSONObject jobject,
      final WritableMap map,
      final PackerContext packerContext)
      throws JSONException, IOException {
    final String type = jobject.getString("type");
    final Unpacker u = mUnpackerMap.get(type);
    if (u == null) {
      throw new IllegalStateException("Unknown type for unpacker: '" + type + "'");
    }
    u.unpack(ivalue, jobject, map, packerContext);
  }
}

File: react-native-pytorch-core/ios/ML/Processing/PackerRegistry.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import SwiftyJSON

enum PackerRegistryError: Error {
    case packerNameAlreadyRegistered
    case unpackerNameAlreadyRegistered
}

public class PackerRegistry {

    enum PackerRegistryError: Error {
        case packerNameAlreadyRegistered
        case unpackerNameAlreadyRegistered
        case unknownPackerType
        case unknownUnpackerType
    }

    private static var mPackerMap = [String: Packer]()
    private static var mUnpackerMap = [String: Unpacker]()

    public static func register(type: String, packer: Packer) throws {
        if mPackerMap.keys.contains(type) {
            throw PackerRegistryError.packerNameAlreadyRegistered
        }

        mPackerMap[type] = packer
    }

    public static func register(type: String, unpacker: Unpacker) throws {
        if mUnpackerMap.keys.contains(type) {
            throw PackerRegistryError.unpackerNameAlreadyRegistered
        }

        mUnpackerMap[type] = unpacker
    }

    public static func pack(
        modelSpec: JSON,
        params: NSDictionary,
        packerContext: PackerContext)
    throws ->  IValue? {
        let packSpec = modelSpec["pack"]
        guard let type: String = packSpec["type"].string else {
            throw BaseIValuePackerError.invalidPackType
        }

        guard let packer: Packer = mPackerMap[type] else {
            throw PackerRegistryError.unknownPackerType
        }

        return try packer.pack(modelSpec: modelSpec, params: params, packerContext: packerContext)
    }

    public static func unpack(
        ivalue: IValue,
        modelSpec: JSON,
        result: inout [String: Any],
        packerContext: PackerContext)
    throws {
        let unpackSpec = modelSpec["unpack"]
        guard let type: String = unpackSpec["type"].string else {
            throw BaseIValuePackerError.invalidUnpackType
        }

        guard let unpacker: Unpacker = mUnpackerMap[type] else {
            throw PackerRegistryError.unknownUnpackerType
        }

        try unpacker.unpack(ivalue: ivalue, modelSpec: modelSpec, result: &result, packerContext: packerContext)
    }
}
